### PR TITLE
docs(admin-server): admin api + websockets + json rpc

### PR DESCRIPTION
**Pull request by frdomovic**
# docs(admin-server): admin api + websockets + json rpc #474

## Summary:
Added README.md docs for admin server which is split into three parts: admin-api, json rpc and websocket

### CODE REVIEW

File: crates/server/README.md [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+
+**Application Management**: The API provides functionalities to install new applications and start contexts. This enables administrators to dynamically manage the software running on the node.
+
+**Key Management**: Administrators can manage root and client keys through the API, ensuring secure access and control over the node.
+
+**Authentication**: The Admin API facilitates user authentication via selected wallets, currently supporting MetaMask and NEAR networks. Authentication details will be explained in later sections.
+
+**Integration with Web Applications**: The authentication mechanism is also used by web applications designed to interact with P2P applications installed on the node, ensuring secure and authenticated access.
+
+### 2. Admin JSON rpc
```
Comment (miraclx) [inline crates/server/README.md]: this is not admin

File: crates/server/README.md [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+
+#### Query Method
+The `Query` method retrieves data from the applications running on the Calimero nodes. For instance, in the Only Peers forum application, posts and comments stored in the application's storage can be queried using the JSON-RPC interface. This enables users to fetch and display content from the forum.
+
+#### Mutate Method
+The `Mutate` method allows modification of the application's data. For example, in the Only Peers forum application, users can create new posts or comments. The Mutate method updates the application's storage with these new entries, facilitating dynamic content creation and interaction within the application.
+
+The Admin JSON-RPC component is crucial for enabling interactive and real-time data management between P2P application front-ends and the back-end contexts running on Calimero nodes.
+
+### 3. Admin Websocket
```
Comment (miraclx) [inline crates/server/README.md]: nor is this

File: crates/server/README.md [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+## Calimero Admin Workflows
+
+### Workflow for Admin Dashboard
+
+```mermaid
+flowchart TD
+    A[Admin Dashboard] -->|HTTP Requests| B[Admin API]
+    B -->|HTTP Responses| A
+    B -->|Query/Mutate| C[RocksDB]
```
Comment (miraclx) [inline crates/server/README.md]: RocksDB is an implementation detail of much lower-level pieces, it shouldn't need to come up here

File: crates/server/README.md [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+```mermaid
+flowchart TD
+    A[P2P Web Application] -->|HTTP Auth Request| B[Admin API]
+    B -->|Auth Response| A
+    A -->|JSON-RPC Query/Mutate| C[Admin JSON rpc]
+    C -->|Query/Mutate| D[Node]
+    D -->|Response| C
+    C -->|Response| A
+
+    subgraph Admin Server
```
Comment (miraclx) [inline crates/server/README.md]: just `Server` will suffice

File: crates/server/README.md (hunk 1 above)
Comment (frdomovic) [reply crates/server/README.md]: updated

File: crates/server/README.md (hunk 2 above)
Comment (frdomovic) [reply crates/server/README.md]: updated

File: crates/server/README.md (hunk 4 above)
Comment (frdomovic) [reply crates/server/README.md]: updated

File: crates/server/README.md (hunk 3 above)
Comment (frdomovic) [reply crates/server/README.md]: removed

File: crates/server/README.md [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+    B -->|HTTP Responses| A
+    B -->|Node Communication| D[Node]
+    D -->|Response| B
+
+    subgraph Server
+        B
+    end
+```
+
+### Workflow for P2P Web applications and JSON rpc
```
Comment (MatejVukosav) [inline crates/server/README.md]: Not really useful imo as there is no clear order of actions. Use a sequence diagram for this kind of graphs

File: crates/server/README.md [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+
+## Calimero Server Workflows
+
+### Workflow for Admin Dashboard
+
+```mermaid
+flowchart TD
+    A[Admin Dashboard] -->|HTTP Requests| B[Admin API]
+    B -->|HTTP Responses| A
+    B -->|Node Communication| D[Node]
```
Comment (MatejVukosav) [inline crates/server/README.md]: Basic Client -> Server communication. 

Node acts as backend implementation. Not sure this graph is needed.

File: crates/server/README.md [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+
+## Admin API endpoints
+
+The Admin API endpoints are split into protected and unprotected routes, where protected routes require authentication.
+
+**Base path**: `/admin-api`
+
+### Protected Routes
+
+These routes require authentication and are protected by the AuthSignatureLayer.
```
Comment (MatejVukosav) [inline crates/server/README.md]: `AuthSignatureLayer` is not common for people to know what it is. 

`These routes require authentication using auth headers`. Then describe how it looks like and how to generate one.

File: crates/server/README.md [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+
+**11. List Contexts**
+ - **Path**: `/contexts`
+ - **Method**: `GET`
+ - **Description**: Lists all contexts.
+
+**12. Delete Auth Keys**
+ - **Path**: `/identity/keys`
+ - **Method**: `DELETE`
+ - **Description**: Deletes authentication keys.
```
Comment (MatejVukosav) [inline crates/server/README.md]: `Deletes all root and client keys`

File: crates/server/README.md [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+**6. List Dev Applications**
+
+ - **Path**: `/dev/applications`
+ - **Method**: `GET`
+ - **Description**: Lists all development applications.
+
+
+## JSON rpc endpoints
+
+The JSON-RPC server endpoints are structured to handle various request types and are configured based on a JSON-RPC configuration that determines if the server is enabled
```
Comment (MatejVukosav) [inline crates/server/README.md]: Do we have more than one endpoint?

Comment (MatejVukosav) [inline crates/server/README.md]: `are configured based on a JSON-RPC configuration that determines if the server is enabled` - I don't get this

File: crates/server/README.md [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+
+**Base path**: `/jsonrpc`
+
+**1. Handle JSON-RPC Request**
+
+- **Path**: `/jsonrpc`
+- **Method**: `POST`
+- **Description**: Handles incoming JSON-RPC requests, which can be `query` or `mutate` requests, processes them, and returns the appropriate response.
```
Comment (MatejVukosav) [inline crates/server/README.md:241]: Explain what are query and mutate requests. People not familiar with GraphQL wouldn't understand it

File: crates/server/README.md [hunk 11]
```diff
@@ -0,0 +1,257 @@
+# Calimero Server
```
Comment (MatejVukosav) [inline crates/server/README.md]: Let's not use `Calimero Server` as it indicates that the company exposed some central server.

Call it `Node Server` until we define a better name.

File: crates/server/README.md [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+- [Admin API endpoints](#admin-api-endpoints)
+    - [Protected Routes](#protected-routes)
+    - [Unprotected Routes](#unprotected-routes)
+- [JSON rpc endpoints](#json-rpc-endpoints)
+- [Websocket endpoints](#websocket-endpoints)
+- [Examples](#examples)
+
+## Introduction
+
+The Calimero Server is a server component of the Calimero Node that has crucial part in
```
Comment (MatejVukosav) [inline crates/server/README.md]: Node server is a component in node that facilitates node administration and enables communication with the logic of application (loaded wasm) in participating contexts.

File: crates/server/README.md [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+managing the Calimero Admin Dashboard and the P2P Applications built for Calimero Nodes.
+
+Calimero Server component is split into 3 parts:
+
+### 1. Admin API
+
+The Admin API component of the Calimero Server connects web applications with the node and its functionalities. It is primarily utilized by the Admin Dashboard to query and manage various aspects of the node, including:
```
Comment (MatejVukosav) [inline crates/server/README.md]: Doesn't need to be a web application. It exposes API which can be consumed

File: crates/server/README.md [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+
+The JSON-RPC component of the Calimero Server facilitates communication between the front-end web applications of P2P systems and the application binaries (contexts) running on Calimero nodes. This allows seamless interaction and data management for applications.
+
+The JSON-RPC interface provides two primary methods:
+
+- Query
+- Mutate
+
+#### Query Method
+The `Query` method retrieves data from the applications running on the Calimero nodes. For instance, in the Only Peers forum application, posts and comments stored in the application's storage can be queried using the JSON-RPC interface. This enables users to fetch and display content from the forum.
```
Comment (MatejVukosav) [inline crates/server/README.md]:
```suggestion
The `Query` method retrieves data from the applications running on the node. For instance, in the Only Peers forum application, posts and comments stored in the application's storage can be queried using the JSON-RPC interface. This enables users to fetch and display content from the forum.
```

File: crates/server/README.md [hunk 15]
```diff
@@ -0,0 +1,257 @@
...
+
+#### Mutate Method
+The `Mutate` method allows modification of the application's data. For example, in the Only Peers forum application, users can create new posts or comments. The Mutate method updates the application's storage with these new entries, facilitating dynamic content creation and interaction within the application.
+
+The JSON-RPC component is crucial for enabling interactive and real-time data management between P2P application front-ends and the back-end contexts running on Calimero nodes.
```
Comment (MatejVukosav) [inline crates/server/README.md]: `real-time data management` is enabled with functionality below

Comment (MatejVukosav) [inline crates/server/README.md]: `front-ends and the back-end contexts` <- we don't have front-ends and back-end contexts.

File: crates/server/README.md [hunk 16]
```diff
@@ -0,0 +1,257 @@
...
+
+### 3. Websocket
+
+The WebSocket is used for subscribing to and unsubscribing from certain contexts within the Calimero server. Defined handlers manage subscription states for WebSocket connections, allowing clients to receive updates about specific contexts they are interested in.
```
Comment (MatejVukosav) [inline crates/server/README.md]: Above JSON-RPC enables interaction with the application, and here WebSocket enable interaction with the context.

File: crates/server/README.md [hunk 17]
```diff
@@ -0,0 +1,257 @@
...
+
+**1. Handle WebSocket Request**
+
+ - **Path**: `/ws`
+ - **Method**: `GET`
+ - **Description**: Handles incoming WebSocket requests, which can be subscribe or unsubscribe requests, processes them, and returns the appropriate response.
+
+
+## Examples
+Examples of Admin Server usage can be found within the [Admin Dashboard](https://github.com/calimero-network/admin-dashboard) and the [Only Peers example](https://github.com/calimero-network/only-peers-client
```
Comment (MatejVukosav) [inline crates/server/README.md]: All communication with the node should be exposed through `calimero sdk`.

Comment (frdomovic) [reply crates/server/README.md]: Updated

File: crates/server/README.md (hunk 14 above)
Comment (frdomovic) [reply crates/server/README.md]: Updated

File: crates/server/README.md (hunk 13 above)
Comment (frdomovic) [reply crates/server/README.md]: Updated

File: crates/server/README.md (hunk 12 above)
Comment (frdomovic) [reply crates/server/README.md]: Updated

File: crates/server/README.md (hunk 11 above)
Comment (frdomovic) [reply crates/server/README.md]: Updated

File: crates/server/README.md (hunk 9 above)
Comment (frdomovic) [reply crates/server/README.md]: Removed this part

Comment (frdomovic) [reply crates/server/README.md]: Updated to "endpoint" as it has only 1 endpoint that parses requests later in the code

File: crates/server/README.md (hunk 8 above)
Comment (frdomovic) [reply crates/server/README.md]: Updated

File: crates/server/README.md (hunk 7 above)
Comment (frdomovic) [reply crates/server/README.md]: Updated and added explanations for auth headers

File: crates/server/README.md (hunk 15 above)
Comment (frdomovic) [reply crates/server/README.md]: Removed

Comment (frdomovic) [reply crates/server/README.md]: Removed

File: crates/server/README.md (hunk 6 above)
Comment (frdomovic) [reply crates/server/README.md]: Removed graphs

File: crates/server/README.md [hunk 18]
```diff
@@ -0,0 +1,214 @@
...
+
+**Key Management**: Administrators can manage root and client keys through the API, ensuring secure access and control over the node.
+
+**Authentication**: The Admin API facilitates user authentication via selected wallets, currently supporting MetaMask and NEAR networks. Authentication details will be explained in later sections.
+
+**Integration with Web Applications**: The authentication mechanism is also used by web applications designed for applications (loaded wasm) in participating context, ensuring secure and authenticated access.
+
+### 2. JSON rpc
+
+The JSON-RPC component of the Node Server facilitates communication between the clients and the application binaries (contexts) running on node. This allows seamless interaction and data management for applications.
```
Comment (MatejVukosav) [inline crates/server/README.md]: `application binaries (contexts)` - context is more that just application binary


> communication between the client and the context.

File: crates/server/README.md [hunk 19]
```diff
@@ -0,0 +1,214 @@
...
+
+These routes require authentication using auth headers.
+Auth headers are generated using `createAuthHeader` function from the `calimero sdk` library.
+
+Parts of the Auth Headers
+ 1. `wallet_type`: Specifies the type of wallet used (e.g., NEAR).
+ 2. `signing_key`: Encoded public key used for signing the request.
+ 3. `signature`: Encoded signature generated from the payload hash.
+ 4. `challenge`: Encoded hash of the payload, serving as a challenge.
+ 5. `context_id`: Optional context identifier for additional request context.
```
Comment (MatejVukosav) [inline crates/server/README.md:88]: It is optional only for admin dashboard right? For all other apps it is mandatory

File: crates/server/README.md (hunk 18 above)
Comment (frdomovic) [reply crates/server/README.md]: Updated

File: crates/server/README.md (hunk 5 above)
Comment (frdomovic) [reply crates/server/README.md]: Added sequence diagrams for all. Do we also need workflow for admin dashboard actions?

**Review by MatejVukosav — APPROVED** [review]

File: crates/server/README.md (hunk 5 above)
Comment (MatejVukosav) [reply crates/server/README.md]: It is good for now.

File: crates/server/README.md [hunk 20]
```diff
@@ -0,0 +1,291 @@
+# Node Server
+
+- [Introduction](#introduction)
+    - [Admin API](#1admin-api)
+    - [JSON rpc](#2json-rpc)
+    - [Websocket](#3websocket)
```
Comment (miraclx) [inline crates/server/README.md]: broken links?

File: crates/server/README.md [hunk 21]
```diff
@@ -0,0 +1,291 @@
...
+- [Websocket endpoints](#websocket-endpoints)
+- [Examples](#examples)
+
+## Introduction
+
+Node Server is a component in node that facilitates node administration and enables communication with the logic of an application (loaded wasm) in participating contexts.
+
+Node Server component is split into 3 parts:
+ 1. [Admin API](https://github.com/calimero-network/core/blob/feat-admin_api_docs/crates/server/src/admin/service.rs)
+ 2. [JSON rpc](https://github.com/calimero-network/core/blob/feat-admin_api_docs/crates/server/src/jsonrpc.rs)
```
Comment (miraclx) [inline crates/server/README.md:24]: Also, can we fix the casing of "RPC"

In some places, it's "rpc" in other's it's "RPC"

File: crates/server/README.md (hunk 20 above)
Comment (frdomovic) [reply crates/server/README.md]: Updated

File: crates/server/README.md (hunk 21 above)
Comment (frdomovic) [reply crates/server/README.md:24]: renamed to "rpc"

**Pull request merged**